windows/src/engine/keyman32/tests/keyman-engine-tests/kmprocessactionstests.cpp

```diff
@@ -0,0 +1,168 @@
+
+  UninitialiseProcess(FALSE);
+  Globals_UninitProcess();
+}
+
+// KM_KBP_IT_MARKER - processMarker Deadkey
+TEST(AITIP, processMarkertest) {
+  Globals_InitProcess();
+
+  WCHAR callbuf[MAXCONTEXT];
+  AITIP testApp;
+  WCHAR expectedContext[] = {UC_SENTINEL, CODE_DEADKEY, 2, 0};
+  uintptr_t marker               = 2;
+  km_kbp_action_item itemAddMarker = {KM_KBP_IT_MARKER, {0,}, {marker}};
+
+  processMarker(&testApp, &itemAddMarker);
+  WCHAR *contextBuf = testApp.ContextBufMax(MAXCONTEXT);
+  EXPECT_STREQ(contextBuf, expectedContext);
+
+  UninitialiseProcess(FALSE);
+  Globals_UninitProcess();
+}
+
+// KM_KBP_IT_BACK - processBack 
+// First test processing a backspace for a deadkey
+TEST(AITIP, processBackDeadkeytest) {
+  Globals_InitProcess();
+  Globals_InitThread();
+
+  WCHAR callbuf[MAXCONTEXT];
+  AITIP testApp;
+  WCHAR expectedContext[] = {'A', 0};
+  km_kbp_action_item itemAddChar = {KM_KBP_IT_CHAR, {0,}, {'A'}};
+
+  processUnicodeChar(&testApp, &itemAddChar);
+  uintptr_t marker                 = 2;
+  km_kbp_action_item itemAddMarker = {KM_KBP_IT_MARKER, {0,}, {marker}};
+  processMarker(&testApp, &itemAddMarker);
+  km_kbp_action_item itemBackSpace = {KM_KBP_IT_BACK};
+  itemBackSpace.backspace.expected_type  = KM_KBP_IT_MARKER;
+  itemBackSpace.backspace.expected_value = marker;
+  processBack(&testApp, &itemBackSpace);
+  WCHAR *contextBuf = testApp.ContextBufMax(MAXCONTEXT);
+  EXPECT_STREQ(contextBuf, expectedContext);
+
+  UninitialiseProcess(FALSE);
+  Globals_UninitProcess();
+}
+
+// KM_KBP_IT_BACK - processBack
+// Press Backspace for a normal character
+// Also test for Unknown Character
+TEST(AITIP, processBackCharactertest) {
+  Globals_InitProcess();
+  Globals_InitThread();
+
+  WCHAR callbuf[MAXCONTEXT];
+  AITIP testApp;
+  WCHAR expectedContext[] = {'A', 0};
+  WCHAR expectedContextFinal[] = {0};
+  km_kbp_action_item itemAddChar = {KM_KBP_IT_CHAR, {0,}, {'A'}};
+
+  processUnicodeChar(&testApp, &itemAddChar);
+  itemAddChar.character            = 'B';
+  processUnicodeChar(&testApp, &itemAddChar);
+  km_kbp_action_item itemBackSpace       = {KM_KBP_IT_BACK};
+  itemBackSpace.backspace.expected_type  = KM_KBP_IT_CHAR;
+  itemBackSpace.backspace.expected_value = 'B';
+  // backspace
+  processBack(&testApp, &itemBackSpace);
+  WCHAR *contextBuf = testApp.ContextBufMax(MAXCONTEXT);
+  EXPECT_STREQ(contextBuf, expectedContext);
+  // backspace for unknown it should backspace a character
+  itemBackSpace.type = KM_KBP_BT_UNKNOWN;
+  processBack(&testApp, &itemBackSpace);
+  contextBuf = testApp.ContextBufMax(MAXCONTEXT);
+  EXPECT_STREQ(contextBuf, expectedContextFinal);
+
+  UninitialiseProcess(FALSE);
+  Globals_UninitProcess();
+}
+
+// KM_KBP_IT_BACK - processBack
+// Press Backspace for a character doesn't match expected character
+// Note currently we don't check for a character match this should be updated
+TEST(AITIP, processBackUnexpectedChartest) {
+  Globals_InitProcess();
+  Globals_InitThread();
```

Comment:
Is there any way we can move these into a setup function and the corresponding cleanup into a teardown function so we are not repeating for each test?

Reply:
Looks like I can use a test fixture to do this.